### PR TITLE
Add `crossword` sample

### PR DESCRIPTION
**Pull request by domesticmouse**
Fixes: https://github.com/flutter/flutter/issues/149286

## Pre-launch Checklist

- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I read the [Contributors Guide].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-devrel channel on [Discord].

[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Contributors Guide]: https://github.com/flutter/games/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: samples/crossword/lib/model.dart [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+  Location get location;
+
+  /// The direction of this word in the crossword.
+  Direction get direction;
+
+  /// Compare two CrosswordWord by coordinates, x then y.
+  static int locationComparator(CrosswordWord a, CrosswordWord b) {
+    final compareRows = a.location.y.compareTo(b.location.y);
+    final compareColumns = a.location.x.compareTo(b.location.x);
+    return switch (compareColumns) { 0 => compareRows, _ => compareColumns };
```
Comment (filiph) [inline samples/crossword/lib/model.dart:93]: [nit] A little too fancy for my taste, but that's okay. I think this should have a trailing comma so that the cases are below each other.

File: samples/crossword/lib/model.dart [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+/// A character in a crossword. This is a single character at a location in a
+/// crossword. It may be part of an across word, a down word, both, but not
+/// neither. The neither constraint is enforced elsewhere.
+abstract class CrosswordCharacter
+    implements Built<CrosswordCharacter, CrosswordCharacterBuilder> {
+  static Serializer<CrosswordCharacter> get serializer =>
+      _$crosswordCharacterSerializer;
+
+  /// The character at this location.
+  String get character;
```
Comment (filiph) [inline samples/crossword/lib/model.dart:124]: [nit] I'd love to see `package:characters` used here. I assume this is out of scope of this PR, though.

Comment (filiph) [reply samples/crossword/lib/model.dart:124]: Actually, ignore me. I see that the rest of the code assumes English (including words.txt).

File: samples/crossword/lib/utils.dart [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    final minutes = inMinutes.remainder(60).toString().padLeft(2, '0');
+    final seconds = inSeconds.remainder(60).toString().padLeft(2, '0');
+    return switch ((inDays, inHours, inMinutes, inSeconds)) {
+      (0, 0, 0, _) => '${inSeconds}s',
+      (0, 0, _, _) => '$inMinutes:$seconds',
+      (0, _, _, _) => '$inHours:$minutes:$seconds',
+      _ => '$inDays days, $hours:$minutes:$seconds',
+    };
+  }
+}
```
Comment (filiph) [inline samples/crossword/lib/utils.dart:34]: This is very nice and I'm probably going to steal this.

File: samples/crossword/lib/widgets/crossword_generator_widget.dart [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+              ),
+            ),
+          );
+
+          if (character != null) {
+            return AnimatedContainer(
+              duration: Durations.extralong1,
+              curve: Curves.easeInOut,
+              color: explorationCell
+                  ? Theme.of(context).colorScheme.primary
```
Comment (filiph) [inline samples/crossword/lib/widgets/crossword_generator_widget.dart:60]: [nit] Lots of `Theme.of(context).colorScheme`. Could be extracted to a local variable at the top of the function: both for readability and (tiny) perf improvement.

File: samples/crossword/test/model_test.dart [hunk 5]
```diff
@@ -0,0 +1,782 @@
...
+      words: [thisWord, thatWord],
+    );
+
+    expect(crossword.words.isNotEmpty, true);
+    expect(crossword.words.length, 2);
+    expect(
+        crossword.words
+            .rebuild(
+              (b) => b.where((b) => b.direction == Direction.across),
+            )
```
Comment (filiph) [inline samples/crossword/test/model_test.dart:46]: Does this really need a rebuild step? Is there no `where` on built lists? Or at least `toList().where()`?

**Review by filiph — APPROVED** [review]
LGTM with some random comments.

File: samples/crossword/lib/model.dart (hunk 2 above)
Comment (domesticmouse) [reply samples/crossword/lib/model.dart:124]: How would I use `package:characters` here? I was aiming for this to be usable with non-ascii word sets. What did I miss?

File: samples/crossword/lib/model.dart (hunk 1 above)
Comment (domesticmouse) [reply samples/crossword/lib/model.dart]: Done

File: samples/crossword/lib/model.dart (hunk 2 above)
Comment (filiph) [reply samples/crossword/lib/model.dart:127]: Not that I'm 100% sure, but I'd say that words (and their characters) should be, _ideally_, represented by the `Characters` class. https://pub.dev/packages/characters It's a little harder to work with than `String`.

To be honest, I don't think this is a priority to use Characters here, and possibly it's a complete waste of your time. As far as I know, most (or all) human languages will behave as expected even if you keep String here. I just tried:

```dart
void main() {
  print('こんにちは世界！'.length);
}
```

It prints 8, which is correct.

`Characters` make sense when you deal with strings such as "Hello, 🌎!" But that's probably impossible to encounter in crosswords puzzles, so you're good.

Comment (domesticmouse) [reply samples/crossword/lib/model.dart:127]: I'd need to convert a `String` to a `List<String>`, which is effectively what a `Characters` effectively is. I thought I did that somewhere, but it's been a month or so since I was really in this codebase.

Comment (filiph) [reply samples/crossword/lib/model.dart:127]: Yeah, I really think you're good with using `String`. I commented before thinking.

File: samples/crossword/test/model_test.dart (hunk 5 above)
Comment (domesticmouse) [reply samples/crossword/test/model_test.dart:46]: I got into the habit of always rebuilding. To save the `.toBuiltList()` call.

File: samples/crossword/lib/widgets/crossword_generator_widget.dart (hunk 4 above)
Comment (domesticmouse) [reply samples/crossword/lib/widgets/crossword_generator_widget.dart]: done
